gulp-tasks/transpile-typescript.js

```diff
@@ -123,26 +63,87 @@ const pendingChangesMap = {};
  */
-const needsTranspile = (packageName) => {
+function needsTranspile(packageName) {
   return pendingChangesMap[packageName] === true;
-};
+}
+
+/**
+ * Transpiles a package iff it has TypeScript source files.
+ *
+ * @param {string} packagePath
+ * @param {Object} [options]
+ */
+async function transpilePackageOrSkip(packagePath, options) {
+  // `packagePath` will be posix style because it comes from `glob()`.
+  const packageName = packagePath.split('/').slice(-1)[0];
 
-gulp.task('transpile-typescript', gulp.series(packageRunnner(
-    'transpile-typescript', 'all', transpilePackageOrSkip)));
+  if (await fse.pathExists(upath.join(packagePath, 'tsconfig.json'))) {
+    await queueTranspile(packageName, options);
+  } else {
+    logHelper.log(ol`Skipping package '${packageName}', which has
+        not yet been converted to typescript.`);
+  }
+}
+
+/**
+ * Transpiles a package's source TypeScript files to `.js` and `.mjs` files
+ * in the root package directory, along with the corresponding `.d.ts` files.
+ *
+ * @param {string} packageName
+ * @param {Object} [options]
+ * @param {boolean} [options.failOnError=true]
+ */
+async function transpilePackage(packageName, {failOnError = true} = {}) {
+  try {
+    // Compile TypeScript for the given project.
+    // Reference the local `node_modules` version of `tsc` since on Windows
+    // it will call the version in `Microsoft SDKs/TypeScript`.
+    await execa('tsc', ['-b', `packages/${packageName}`], {preferLocal: true});
+
+    // Mirror all `.ts` files with `.mjs` files.
+    const tsFiles = await globby(`**/*.ts`, {
+      cwd: upath.join('packages', packageName, 'src'),
+    });
+
+    for (const tsFile of tsFiles) {
+      const assetBasename = upath.basename(tsFile, '.ts');
+      const mjsFile = upath.join('packages',
+          packageName, upath.dirname(tsFile), `${assetBasename}.mjs`);
+
+      if (!(await fse.pathExists(mjsFile))) {
+        const mjsSource = `export * from './${assetBasename}.js';`;
+
+        await fse.outputFile(mjsFile, mjsSource);
+      }
+    }
+  } catch (error) {
+    if (failOnError) {
+      throw error;
+    }
+
+    logHelper.error(error.stdout);
+  }
+}
 
-gulp.task('transpile-typescript:watch', () => {
-  const watcher = gulp.watch(`./${global.packageOrStar}/workbox-*/src/**/*.ts`);
+function transpile_typescript_watch() {
+  const watcher = watch(`./${global.packageOrStar}/workbox-*/src/**/*.ts`);
   watcher.on('all', async (event, file) => {
-    const changedPkg = path.relative(packagesDir, file).split(path.sep)[0];
+    const changedPkg = upath.relative(packagesDir, file).split(upath.sep)[0];
 
     pendingChangesMap[changedPkg] = true;
     await queueTranspile(changedPkg, {failOnError: false});
     pendingChangesMap[changedPkg] = false;
   });
-});
+}
 
 module.exports = {
-  transpilePackageOrSkip,
-  queueTranspile,
-  needsTranspile,
+  // These are used as functions from other modules, not as gulp tasks.
+  functions: {
```

Comment:
Would it make sense for these functions to live in a separate file, and then this file and the external consumers could just pull in that?

Reply:
I was initially going to do that, but after making that refactor, there was some cross-dependencies that made it too awkward to break things up.